File: workers/_shared/token_plan.js
/**
 * Lecture du PLAN porté par un token vérifié (claims sv ≥ 3).
 *
 * POURQUOI CE MODULE. Depuis le schéma 3, le passe transporte lui-même le plan
 * choisi sur le site (`p`), le consentement au corpus vocal (`cc`) et la version
 * des textes légaux acceptés (`cv`). Ces trois claims sont dans les octets
 * signés : elles font autorité, contrairement aux en-têtes envoyés par le
 * client, qui sont déclaratifs et forgeables. Chaque worker qui décide du sort
 * d'un enregistrement vocal doit lire ICI, et nulle part ailleurs.
 *
 * CONTRAT. `readPlan(claims)` ne prend QUE des claims déjà vérifiées par
 * `verifyToken()` (`v.claims`). Il ne vérifie aucune signature et n'en a pas la
 * charge ; lui passer un payload non vérifié serait une faille.
 *
 * Renvoie { sv, plan, corpusConsent, legalVersion } :
 *   - sv            : version de schéma lue (0 si illisible)
 *   - plan          : 'free' | 'paid' | null (null = pas de plan porté → repli
 *                     historique : écran de consentement in-app)
 *   - corpusConsent : booléen — TOUJOURS false hors du plan 'free'
 *   - legalVersion  : chaîne non vide, ou null
 *
 * DEUX RÈGLES DURES, volontairement redondantes avec `token_verify.js` :
 *   1. sv < 3 ⇒ plan null. Un ancien passe ne consent à rien par défaut.
 *   2. plan 'paid' ⇒ corpusConsent false quoi que dise `cc`. Le passe Payant est
 *      l'alternative SANS enregistrement : c'est ce qui rend libre le
 *      consentement du plan Gratuit (RGPD art. 7(4)). Un 'paid' avec `cc: true`
 *      est une incohérence — le tokeniser la refuse à l'émission ; si elle
 *      arrivait ici malgré tout, elle ne doit jamais valoir autorisation.
 */

// Première version de schéma qui porte les claims de plan.
const SCHEMA_VERSION_PLAN = 3;

// Plans admis pour la claim `p`.
const TOKEN_PLANS = new Set(['free', 'paid']);

/** Résultat neutre : aucun plan porté, aucun consentement, aucune version. */
function aucunPlan(sv) {
  return { sv, plan: null, corpusConsent: false, legalVersion: null };
}

/**
 * Lit le plan d'un jeu de claims VÉRIFIÉES.
 * @param {object|null|undefined} claims — `v.claims` de `verifyToken()`.
 */
export function readPlan(claims) {
  if (!claims || typeof claims !== 'object' || Array.isArray(claims)) {
    return aucunPlan(0);
  }

  const sv = Number.isInteger(claims.sv) ? claims.sv : 0;
  if (sv < SCHEMA_VERSION_PLAN) return aucunPlan(sv);

  // Forme strictement identique à celle exigée par `verifyToken()` : un claims
  // hors forme retombe sur le repli historique plutôt que sur une supposition.
  if (!TOKEN_PLANS.has(claims.p)) return aucunPlan(sv);

  const plan = claims.p;
  const legalVersion =
    typeof claims.cv === 'string' && claims.cv.length > 0 ? claims.cv : null;
  // `cc` non booléen ⇒ pas de consentement ; 'paid' ⇒ jamais de consentement.
  const corpusConsent = plan === 'free' && claims.cc === true;

  return { sv, plan, corpusConsent, legalVersion };
}

File: lib/core/services/token_plan.dart
// lib/core/services/token_plan.dart
//
// Le PLAN porté par le token (`sv: 3`) et ce qui l'accompagne.
//
// Depuis le 2026-09-02, le passe est créé sur mental-et.com/inscription et
// porte le choix de l'utilisateur : passe Gratuit (le bilan est financé par un
// enregistrement vocal) ou passe Payant (même bilan, aucun enregistrement).
// Le consentement au corpus vocal est recueilli SUR LE SITE, avant l'émission,
// et voyage dans le token signé : c'est lui la preuve (art. 7 RGPD), pas un
// écran in-app.
//
// Miroir Dart de `readPlan()` dans workers/_shared/token_plan.js.
//
// ⚠️ Un `TokenPlanInfo` ne vaut que s'il a été dérivé de claims dont la
//    SIGNATURE a été vérifiée (ou d'un token DEV `M2.` en debug). Voir
//    `TokenClaimsReader.planFromToken` : le payload non vérifié n'est jamais
//    une source de plan.

/// Plan porté par le token.
///
/// [unknown] couvre tout ce qui n'est pas un plan lisible et digne de foi :
/// token `sv: 2` (antérieur au plan), claims de plan mal formés, token absent,
/// signature non vérifiable. Le repli est alors le comportement historique :
/// l'écran de consentement in-app.
enum TokenPlan { free, paid, unknown }

/// Ce que le token dit du plan, du consentement au corpus et du texte légal
/// accepté. Immuable : c'est une lecture, jamais un état modifiable.
class TokenPlanInfo {
  /// Plan choisi à l'inscription.
  final TokenPlan plan;

  /// Consentement au corpus vocal (claim `cc`). Toujours `false` hors du plan
  /// Gratuit : un passe Payant n'enregistre rien, donc ne cède rien.
  final bool corpusConsent;

  /// Version des textes légaux acceptée à l'inscription (claim `cv`), telle
  /// qu'elle doit être consignée comme preuve. `null` hors `sv: 3`.
  final String? legalVersion;

  /// Jour d'émission du passe (claim `d`, jours depuis l'epoch UTC). Sert à
  /// dater le consentement sans jamais lire l'horloge locale — c'est ce qui
  /// rend `ConsentService.syncFromToken` idempotente. `null` si absent.
  final int? issuedDay;

  const TokenPlanInfo({
    required this.plan,
    this.corpusConsent = false,
    this.legalVersion,
    this.issuedDay,
  });

  /// Aucune information exploitable : repli sur l'écran in-app.
  static const TokenPlanInfo unknown = TokenPlanInfo(plan: TokenPlan.unknown);

  /// `true` si le token porte réellement un plan (`sv: 3` bien formé).
  bool get isKnown => plan != TokenPlan.unknown;

  /// `true` si l'étape orale doit être jouée pour ce passe.
  /// Le passe Payant la saute ; un plan inconnu retombe sur l'écran in-app,
  /// qui décide lui-même.
  bool get allowsOralStep => plan != TokenPlan.paid;

  @override
  String toString() => 'TokenPlanInfo(plan: ${plan.name}, '
      'corpusConsent: $corpusConsent, legalVersion: $legalVersion, '
      'issuedDay: $issuedDay)';
}
